Sources/TensorFlow/Operators/Math.swift

```diff
@@ -2224,6 +2227,51 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
             ) / self
         })
     }
+
+    @inlinable
+    func _vjpProduct(squeezingAxes axes: Tensor<Int32>) -> (Tensor, (Tensor) -> Tensor) {
+        /// The gradient can be expressed by dividing the product by each entry of the
+        /// input tensor, but this approach can't deal with zeros in the input.
+        /// Here, we avoid this problem by composing the output as a product of two cumulativeProduct operations.
+        /// Adapted from `_ProdGrad` in Python TensorFlow:
+        /// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py
+        let result = product(squeezingAxes: axes)
+        return (result, { v in
```

Comment:
Currently, the `{ v in ... }` pullback captures all of `self`, which is not efficient.
Ideally, we should add a capture list to capture only the minimal necessary values:
```swift
        return (result, { [rank = self.rank, shape = self.shape, ...] v in
```
However, there's one tricky `self` capture: `let permuted = self.transposed(permutation: perm)`.

I wonder what is the size of `permuted` compared with `self`? If `permuted` is significantly smaller than `self`, then it may be worthwhile to compute it outside of the pullback and to add it to the capture list.

Reply:
What do you mean by 

> If permuted is significantly smaller than self

Isn't it the same Tensor as self, just with permuted dimensions, so they should be equal in the size?

I tried 

```
print("size self", MemoryLayout.size(ofValue: self))  // output: 16
print("size permuted", MemoryLayout.size(ofValue: permuted)) // output: 16
```

Or perhaps I just don`t understand correctly.

Reply:
Oh yes, you're right - `transposed(permutation:)` always returns a tensor with the same size (scalar count).
In that case, let's just capture `self` instead of introducing a capture list.